kafka/main.go

```diff
@@ -54,105 +69,176 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
+		if err != nil {
+			logger.Error().Err(err).Msg("Batching failed")
+			// This should be an app error that needs to communicate if its failure is
+			// temporary or permanent. If temporary we need to handle it and if
+			// permanent we need to commit and move on.
+		}
+	BatchProcessingLoop:
+		for _, msg := range batch {
+			wg.Add(1)
+			if err != nil {
+				// Indicates batch has no more messages. End the loop for
+				// this batch and fetch another.
+				if err == io.EOF {
+					logger.Info().Msg("Batch complete. Ending loop.")
+					break BatchProcessingLoop
 				}
-				logger.Info().Msg(fmt.Sprintf("Processing message for topic %s at offset %d", msg.Topic, msg.Offset))
-				logger.Info().Msg(fmt.Sprintf("Reader Stats: %#v", consumer.Stats()))
-				for _, topicMapping := range topicMappings {
-					if msg.Topic == topicMapping.Topic {
-						go func(
-							msg kafka.Message,
-							topicMapping TopicMapping,
-							providedServer *server.Server,
-							logger *zerolog.Logger,
-						) {
-							err := topicMapping.Processor(
-								msg.Value,
-								topicMapping.ResultProducer,
-								providedServer,
-								logger,
-							)
-							if err != nil {
-								logger.Error().Err(err).Msg("Processing failed.")
-							}
-						}(msg, topicMapping, providedServer, logger)
-						if err := consumer.CommitMessages(ctx, msg); err != nil {
-							logger.Error().Msg(fmt.Sprintf("Failed to commit: %s", err))
+			}
+			logger.Info().Msg(fmt.Sprintf("Processing message for topic %s at offset %d", msg.Topic, msg.Offset))
+			logger.Info().Msg(fmt.Sprintf("Reader Stats: %#v", reader.Stats()))
+			wgDoneDeferred := false
+			// Check if any of the existing topicMappings match the fetched message
+			for _, topicMapping := range topicMappings {
+				if msg.Topic == topicMapping.Topic {
+					wgDoneDeferred = true
+					go func(
+						msg kafka.Message,
+						topicMapping TopicMapping,
+						providedServer *server.Server,
+						logger *zerolog.Logger,
+					) {
+						defer wg.Done()
+						err := topicMapping.Processor(
+							msg,
+							topicMapping.ResultProducer,
+							topicMapping.TolerableEquivalence,
+							providedServer,
+							results,
+							logger,
+						)
+						if err != nil {
+							logger.Error().Err(err).Msg("Processing failed.")
+							results <- err
 						}
+					}(msg, topicMapping, providedServer, logger)
+				}
+			}
+			// If the topic in the message doesn't match andy of the topicMappings
+			// then the goroutine will not be spawned and wg.Done() won't be
+			// called. If this happens, be sure to call it.
+			if !wgDoneDeferred {
+				wg.Done()
+			}
+		}
+		close(results)
+		// Iterate over any failures and get the earliest temporary failure offset
+		var temporaryErrors []*utils.ProcessingError
+		for processingError := range results {
+			if processingError.Temporary {
+				continue
+			} else {
+				temporaryErrors = append(temporaryErrors, processingError)
```

Comment:
I think this `processingError` will be non-temporary, as we check in the corresponding if for Temporary status?